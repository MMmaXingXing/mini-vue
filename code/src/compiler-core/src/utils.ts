import { NodeTypes } from "./ast";

export const isText = (node) => {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;
};
