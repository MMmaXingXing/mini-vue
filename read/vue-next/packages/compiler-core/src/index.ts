export { baseCompile } from './compile'

// Also expose lower level APIs & types
export {
  CompilerOptions,
  ParserOptions,
  TransformOptions,
  CodegenOptions,
  HoistTransform,
  BindingMetadata,
  BindingTypes
} from './options'
export { baseParse, TextModes } from './parse'
export {
  transform,
  TransformContext,
  createTransformContext,
  traverseNode,
  createStructuralDirectiveTransform,
  NodeTransform,
  StructuralDirectiveTransform,
  DirectiveTransform
} from './transform'
export { generate, CodegenContext, CodegenResult } from './codegen'
export {
  ErrorCodes,
  CoreCompilerError,
  CompilerError,
  createCompilerError
} from './errors'

export * from './ast'
export * from './utils'
export * from './babelUtils'
export * from './runtimeHelpers'

export { getBaseTransformPreset, TransformPreset } from './compile'
export { transformModel } from './transforms/vModel'
export { transformOn } from './transforms/vOn'
export { transformBind } from './transforms/vBind'
export { noopDirectiveTransform } from './transforms/noopDirectiveTransform'
export { processIf } from './transforms/vIf'
export { processFor, createForLoopParams } from './transforms/vFor'
export {
  transformExpression,
  processExpression
} from './transforms/transformExpression'
export {
  buildSlots,
  SlotFnBuilder,
  trackVForSlotScopes,
  trackSlotScopes
} from './transforms/vSlot'
export {
  transformElement,
  resolveComponentType,
  buildProps
} from './transforms/transformElement'
export { processSlotOutlet } from './transforms/transformSlotOutlet'
export { generateCodeFrame } from '@vue/shared'

// v2 compat only
export {
  checkCompatEnabled,
  warnDeprecation,
  CompilerDeprecationTypes
} from './compat/compatConfig'
