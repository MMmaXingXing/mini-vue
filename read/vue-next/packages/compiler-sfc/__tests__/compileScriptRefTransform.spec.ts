import { BindingTypes } from '@vue/compiler-core'
import { compileSFCScript as compile, assertCode } from './utils'

// this file only tests integration with SFC - main test case for the ref
// transform can be found in <root>/packages/reactivity-transform/__tests__
describe('sfc ref transform', () => {
  function compileWithReactivityTransform(src: string) {
    return compile(src, { reactivityTransform: true })
  }

  test('$ unwrapping', () => {
    const { content, bindings } = compileWithReactivityTransform(`<script setup>
    import { ref, shallowRef } from 'vue'
    let foo = $(ref())
    let a = $(ref(1))
    let b = $(shallowRef({
      count: 0
    }))
    let c = () => {}
    let d
    </script>`)
    expect(content).not.toMatch(`$(ref())`)
    expect(content).not.toMatch(`$(ref(1))`)
    expect(content).not.toMatch(`$(shallowRef({`)
    expect(content).toMatch(`let foo = (ref())`)
    expect(content).toMatch(`let a = (ref(1))`)
    expect(content).toMatch(`
    let b = (shallowRef({
      count: 0
    }))
    `)
    // normal declarations left untouched
    expect(content).toMatch(`let c = () => {}`)
    expect(content).toMatch(`let d`)
    expect(content).toMatch(`return { foo, a, b, c, d, ref, shallowRef }`)
    assertCode(content)
    expect(bindings).toStrictEqual({
      foo: BindingTypes.SETUP_REF,
      a: BindingTypes.SETUP_REF,
      b: BindingTypes.SETUP_REF,
      c: BindingTypes.SETUP_LET,
      d: BindingTypes.SETUP_LET,
      ref: BindingTypes.SETUP_CONST,
      shallowRef: BindingTypes.SETUP_CONST
    })
  })

  test('$ref & $shallowRef declarations', () => {
    const { content, bindings } = compileWithReactivityTransform(`<script setup>
    let foo = $ref()
    let a = $ref(1)
    let b = $shallowRef({
      count: 0
    })
    let c = () => {}
    let d
    </script>`)
    expect(content).toMatch(
      `import { ref as _ref, shallowRef as _shallowRef } from 'vue'`
    )
    expect(content).not.toMatch(`$ref()`)
    expect(content).not.toMatch(`$ref(1)`)
    expect(content).not.toMatch(`$shallowRef({`)
    expect(content).toMatch(`let foo = _ref()`)
    expect(content).toMatch(`let a = _ref(1)`)
    expect(content).toMatch(`
    let b = _shallowRef({
      count: 0
    })
    `)
    // normal declarations left untouched
    expect(content).toMatch(`let c = () => {}`)
    expect(content).toMatch(`let d`)
    assertCode(content)
    expect(bindings).toStrictEqual({
      foo: BindingTypes.SETUP_REF,
      a: BindingTypes.SETUP_REF,
      b: BindingTypes.SETUP_REF,
      c: BindingTypes.SETUP_LET,
      d: BindingTypes.SETUP_LET
    })
  })

  test('usage in normal <script>', () => {
    const { content } = compileWithReactivityTransform(`<script>
    export default {
      setup() {
        let count = $ref(0)
        const inc = () => count++
        return $$({ count })
      }
    }
    </script>`)
    expect(content).not.toMatch(`$ref(0)`)
    expect(content).toMatch(`import { ref as _ref } from 'vue'`)
    expect(content).toMatch(`let count = _ref(0)`)
    expect(content).toMatch(`count.value++`)
    expect(content).toMatch(`return ({ count })`)
    assertCode(content)
  })

  test('usage /w typescript', () => {
    const { content } = compileWithReactivityTransform(`
      <script setup lang="ts">
        let msg = $ref<string | number>('foo');
        let bar = $ref <string | number>('bar');
      </script>
    `)
    expect(content).toMatch(`import { ref as _ref`)
    expect(content).toMatch(`let msg = _ref<string | number>('foo')`)
    expect(content).toMatch(`let bar = _ref <string | number>('bar')`)
    assertCode(content)
  })

  test('usage with normal <script> + <script setup>', () => {
    const { content, bindings } = compileWithReactivityTransform(`<script>
    let a = $ref(0)
    let c = $ref(0)
    </script>
    <script setup>
    let b = $ref(0)
    let c = 0
    function change() {
      a++
      b++
      c++
    }
    </script>`)
    // should dedupe helper imports
    expect(content).toMatch(`import { ref as _ref } from 'vue'`)

    expect(content).toMatch(`let a = _ref(0)`)
    expect(content).toMatch(`let b = _ref(0)`)

    // root level ref binding declared in <script> should be inherited in <script setup>
    expect(content).toMatch(`a.value++`)
    expect(content).toMatch(`b.value++`)
    // c shadowed
    expect(content).toMatch(`c++`)
    assertCode(content)
    expect(bindings).toStrictEqual({
      a: BindingTypes.SETUP_REF,
      b: BindingTypes.SETUP_REF,
      c: BindingTypes.SETUP_REF,
      change: BindingTypes.SETUP_CONST
    })
  })

  describe('errors', () => {
    test('defineProps/Emit() referencing ref declarations', () => {
      expect(() =>
        compile(
          `<script setup>
        let bar = $ref(1)
        defineProps({
          bar
        })
      </script>`,
          { reactivityTransform: true }
        )
      ).toThrow(`cannot reference locally declared variables`)

      expect(() =>
        compile(
          `<script setup>
        let bar = $ref(1)
        defineEmits({
          bar
        })
      </script>`,
          { reactivityTransform: true }
        )
      ).toThrow(`cannot reference locally declared variables`)
    })
  })
})
