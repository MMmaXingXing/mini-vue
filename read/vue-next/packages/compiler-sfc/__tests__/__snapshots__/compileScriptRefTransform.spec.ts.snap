// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sfc ref transform $ unwrapping 1`] = `
"import { ref, shallowRef } from 'vue'
    
export default {
  setup(__props, { expose }) {
  expose();

    let foo = (ref())
    let a = (ref(1))
    let b = (shallowRef({
      count: 0
    }))
    let c = () => {}
    let d
    
return { foo, a, b, c, d, ref, shallowRef }
}

}"
`;

exports[`sfc ref transform $ref & $shallowRef declarations 1`] = `
"import { ref as _ref, shallowRef as _shallowRef } from 'vue'

export default {
  setup(__props, { expose }) {
  expose();

    let foo = _ref()
    let a = _ref(1)
    let b = _shallowRef({
      count: 0
    })
    let c = () => {}
    let d
    
return { foo, a, b, c, d }
}

}"
`;

exports[`sfc ref transform usage /w typescript 1`] = `
"import { ref as _ref, defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose }) {
  expose();

        let msg = _ref<string | number>('foo');
        let bar = _ref <string | number>('bar');
      
return { msg, bar }
}

})"
`;

exports[`sfc ref transform usage in normal <script> 1`] = `
"import { ref as _ref } from 'vue'

    export default {
      setup() {
        let count = _ref(0)
        const inc = () => count.value++
        return ({ count })
      }
    }
    "
`;

exports[`sfc ref transform usage with normal <script> + <script setup> 1`] = `
"import { ref as _ref } from 'vue'

    let a = _ref(0)
    let c = _ref(0)
    
export default {
  setup(__props, { expose }) {
  expose();

    let b = _ref(0)
    let c = 0
    function change() {
      a.value++
      b.value++
      c++
    }
    
return { a, c, b, change }
}

}"
`;
